core/src/main/scala/org/apache/spark/TaskContextImpl.scala

```diff
@@ -66,8 +66,21 @@ private[spark] class TaskContextImpl(
   /** Marks the task as completed and triggers the listeners. */
   private[spark] def markTaskCompleted(): Unit = {
     completed = true
+    processCallBacks()
+  }
+
+  /** Marks the task for interruption, i.e. cancellation. */
+  private[spark] def markInterrupted(): Unit = {
+    interrupted = true
+    processCallBacks()
```

Comment:
I noticed that the old code did not process callbacks on interruption.  What is the API contract here for markTaskCompleted and markInterrupted()?  Are they mutually-exclusive, e.g. a task can be marked as completed or interrupted, but never both?  I think that we should add assertions to determine if this is the case, since I'm worried that this change might lead to duplicate calling of callbacks if this assumption isn't true.